docs/rfc/rfc-017-abci++-vote-extension-propag.md

```diff
@@ -0,0 +1,568 @@
+
+    In this case, a set of full nodes, denoted *L<sub>h<sub>p</sub></sub>*, fall behind
+    (e.g., temporary disconnection or network partition, memory thrashing, crashes, new nodes)
+    an arbitrary
+    number of heights,
+    say between *h<sub>s</sub>* and *h<sub>p</sub>*, where *h<sub>p</sub>* is the present height.
+
+    Obviously, less than *n<sub>h<sub>p</sub></sub>/3* validators of *h<sub>p</sub>* can be part
+    of *L<sub>h<sub>p</sub></sub>*, since we need enough caught-up validators to make progress
+    at the present height.
+    Further, validator set changes between *h<sub>s</sub>* and *h<sub>p</sub>* impose an
+    additional restriction on which full nodes can be part of *L<sub>h<sub>p</sub></sub>*.
+    Since every node in *L<sub>h<sub>p</sub></sub>* did not take part in any consensus between
+    *h<sub>s</sub>* and *h<sub>p</sub>*, we have this restriction on *L<sub>h<sub>p</sub></sub>*:
+
+    - for each height *h*, where *h<sub>s</sub> ≤ h < h<sub>p</sub>*,
+    | *valset<sub>h</sub>* \ *L<sub>h<sub>p</sub></sub>*  | *> 2n<sub>h<sub>p</sub></sub>/3*
+
+    If this property did not hold for a particular height *h*, where
+    *h<sub>s</sub> ≤ h < h<sub>p</sub>*, Tendermint could not have progressed beyond *h* and
+    therefore *h<sub>p</sub>* could not be the present height (a contradiction).
+
+    These lagging nodes in *L<sub>h<sub>p</sub></sub>* need to catch up. They have to obtain the
+    information needed to make
+    progress from other nodes. For each height *h* between *h<sub>s</sub>* and *h<sub>p</sub>*,
+    this includes the decided block for *h*, and the
+    precommit votes also for *deciding h* (which can be extracted from the block at height *h+1*).
+
+    At a given height  *h<sub>c</sub>* (where possibly *h<sub>c</sub> << h<sub>p</sub>*),
+    a full node in *L<sub>h<sub>p</sub></sub>* will consider itself *caught up*, based on the
+    (maybe out of date) information it is getting from its peers. Then, the node needs to be ready to
+    propose at height *h<sub>c</sub>+1*, which requires having received the vote extensions for
+    *h<sub>c</sub>*.
+    As the vote extensions are *not* stored in the blockchain, and it is difficult to have strong
+    guarantees on *when* a late node considers itself caught up, providing the late note with the right
+    vote extensions for the right height poses a problem.
+
+At this point, we have described and compared all cases raised in discussions leading up to this
+RFC. The list above aims at being exhaustive. The analysis of each case included above makes all of
+them converge into case (h).
+
+### Current Catch-up Mechanisms
+
+We now briefly describe the current catch-up mechanisms in the reactors concerned in Tendermint.
+
+#### Statesync
+
+Full nodes optionally run statesync just after starting, when they start from scratch
+(TODO: folks, please keep me honest here... do we also run statsync when we recover?).
```

Comment:
So, does this mean that, if my node crashes, stays down for a few hours, then recovers ... ?
1. the node will sync with the app via ABCI Info call
2. It will *not* call statesync
3. It will call blocksync to catch up
4. It will switch to consensus when it _thinks_ it is caught up (or after a timeout)

Reply:
It uses the height described there to do light client verification of the blocks at the state height it fetches. That is not the height of the chain to sync to. 

You can see that variable used [here](https://github.com/tendermint/tendermint/blob/da1b871808b352926a24132467ccc5349286d3fa/internal/statesync/reactor.go#L274) for creating a light client, which is then used to [verify light blocks](https://github.com/tendermint/tendermint/blob/da1b871808b352926a24132467ccc5349286d3fa/internal/statesync/stateprovider.go#L143) that it fetches along with the state. The height it syncs to is the highest height known to the process.

Reply:
I see, but can the statesync reactor kick in upon recovery? Or is it only used when starting a node from scratch. That's the part I'll need to adapt in this part of the text if statesync can also be run upon recovery.

Reply:
> Or is it only used when starting a node from scratch.

Currently, it is only used when starting a node up for the first time.
https://github.com/tendermint/tendermint/blob/29e5fbcc648510e4763bd0af0b461aed92c21f30/node/node.go#L287

Reply:
Thanks @williambanfield for confirming. That was my impression when navigating the code but I hadn't spotted this "if"